### PR TITLE
Solo5 platform support

### PR DESCRIPTION
**Pull request by ghost**
This PR makes it possible to run Nim programs as unkernels using the [Solo5](https://github.com/solo5/solo5) platform abstraction. Solo5 was originally developed to abstract MirageOS away from Xen.

What you get is a Nim program that can run as KVM or QEMU guest, or as a strictly seccomp'ed Linux process. Within the sandbox there may be access to virtual block and network devices, but there are no additional "system calls" available and essentially no Libc. The benefit is that if you want to run a Nim server application in a virtualized environment you do that directly without wrapping the application in a sloppy and bloated Linux installation.

Currently there is no file-system available, not until someone writes something like a Nim native FAT or 9P driver. You can serialize data through the block device with something like JSON but I haven't tried it.

Networking does work, but no sockets. I ported a [TAPS](https://datatracker.ietf.org/wg/taps/about/) [library](https://git.sr.ht/~ehmry/nim_taps/commit/solo5) as a sockets replacement and TCP over IPv6 is working. It may be possible to use LwIP sockets inside the unikernel, but I think LwIP has a superior async interface and I've already got that bound to TAPS.

I had to disable most of the content of `std/net` which probably makes this unmergable. I would prefer `std/net` contain things like IP parsing and all the socket and sockaddr stuff be moved to a separate namespace.

I'd be willing to remove Genode platform in exchange for Solo5, because I've ported both Nim and Solo5 to Genode and Solo5 is more general and less maintenance.

### CODE REVIEW

**Comment by Araq** [comment]
We can support Solo5 and Genode at the same time, that's not the issue here. The problem is that we need a better way to have platform specific code. For example how about this:

- net.nim and asyncdispatch.nim use a structure like:

```nim

# net.nim

when defined(genode):
  include net_genode
elif defined(solo5):
  include net_solo5
else:
  include net_default_impl

```

I know that there would be quite some code sharing or code duplication between `net_genode` and `net_default_impl`. The sharing could be done via a `net_common_incl` file and a little bit of code duplication seems preferable over a nested `when` jungle.

- The code generator's `main` generation logic becomes template based and moves to a dedicated module / file.

**In other words, before this PR we need a pure refactoring PR that allows us to target niche platforms more easily.**

**Comment by ghost** [comment]
I'm happy to know this is at least a halfway viable idea.

I see few scenarios:
* Some platforms are supported by Nim but sockets are unavailable and therefore libraries like httpclient are also unavailable. 
* Some platforms are supported that have their own alternatives to sockets and libraries like httpclient have platform `when` switches to select a platform backend.
* Libraries like httpclient use a Nim network abstraction library that hides platform specifics, and might even contain a pure Nim TCP stack.

[Mirage OS](https://mirage.io/) seems to have its own [HTTP library](https://github.com/mirage/ocaml-cohttp#readme) that takes the second approach of multiple backends within the client. Some of these backends are wrappers over platform interfaces but there is also a Mirage TCP/IP stack in OCaml. OCaml is different from Nim in that it doesn't have a standard library with an HTTP client and it also has module parameterization (which I'm not a fan of).

(Genode does have its own network abstraction and its a Plan9 style control file interface. I'd rather Nim interface with Genode on an ethernet level than go one level below the Genode socket abstraction.)

**Comment by ghost** [comment]
FWIW porting higher abstractions to the lwIP API is not that difficult, its the tuning the lwIP compile-time parameters that is tedious. If we commit to supporting socket alternatives then maybe Solo5 could share an lwIP-native backend with FreeRTOS and Zephyr (LwIP exports a low async API that its socket emulation is built on).

**Comment by ghost** [comment]
Now that we're past 2.0.0 I'd like to get back to this.

**Comment by Varriount** [comment]
I'd like to mention, I find the way the Go compiler's source does this to be fairly elegant - have a module for each OS that fulfills an "interface" (in design only) that a higher level module consumes. It prevents heavy nesting and overly long implementations, at the expense of some code duplication.

**Comment by ghost** [comment]
I think we can make elegant separations of OS implementations internally or we can provide elegant abstractions on the surface, but not both. The territory is just really complex and a huge advantage we have over C and Go is native, compile-time `when` conditionals.

**Comment by Araq** [comment]
The only problem with `when` is when it leads to long sections of indented code so that you lose track of where you are so you should use `include` to prevent that from happening.
